088-rebalance-progress-status.md

```diff
@@ -0,0 +1,397 @@
+
+#### State: `Stopped`
+
+Once a rebalance has been stopped, it cannot be resumed.
+However, the `executorState` information from when before the rebalance was stopped may still be of value to users.
+Therefore, we reuse the same value of `executorState` from the previous update.
+
+#### State: `NotReady`
+
+Once a rebalance has been interrupted or completed with errors, it cannot be resumed.
+However, the `executorState` information from when before the rebalance was interrupted or completed may still be of value to users.
+Therefore, we reuse the same value of `executorState` from the previous update.
+
+#### State: `Ready`
+
+The rebalance is complete in this state so there is no data surrounding ongoing partition movement available, so we omit this field in the `progress` section for this state.
+
+### Progress Update Cadence
+
+For ease of implementation and minimizing the load on the CruiseControl REST API server, the operator will only query the `/kafkacruisecontrol/state?substates=executor` endpoint and update the `ConfigMap` upon `KafkaRebalance` resource reconciliation.
+
+In the event that Cruise Control runs into an error when rebalancing, when operator transitions the `KafkaRebalance` resource to the `NotReady` state, it will leave the `progress` section and `ConfigMap` as is.
+In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a "Warning" condition with error message from the failed REST API call but leave the existing progress section and progress `ConfigMap` as is.
+
+When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  ...
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: Rebalancing
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: Warning
+    reason: CruiseControlRestException
+    message: <error_message> [1]  
+  observedGeneration: 1
+  optimizationResult:
+    afterBeforeLoadConfigMap: my-rebalance
+    dataToMoveMB: 0
+    excludedBrokersForLeadership: []
+    excludedBrokersForReplicaMove: []
+    excludedTopics: []
+    intraBrokerDataToMoveMB: 0
+    monitoredPartitionsPercentage: 100
+    numIntraBrokerReplicaMovements: 0
+    numLeaderMovements: 16
+    numReplicaMovements: 0
+    onDemandBalancednessScoreAfter: 95.4347095948149
+    onDemandBalancednessScoreBefore: 89.4347095948149
+    provisionRecommendation: ""
+    provisionStatus: RIGHT_SIZED
+    recentWindows: 1
+  progress:
+    rebalanceProgressConfigMap: my-rebalance-progress
```

Comment:
this can't be `my-rebalance-progress` if we have `afterBeforeLoadConfigMap: my-rebalance` ... are we going to use the same ConfigMap for both, right? So the naming should be consistent in the example.

Reply:
Thank you for the catch, just fixed